File: leetcode/Python/0145-h-binary-tree-postorder.traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

"""
Recursion
"""
class Solution_1:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.postorder(root, res)
        return res

    def postorder(self, node, res):
        if not node:
            return
        self.postorder(node.left, res)
        self.postorder(node.right, res)
        res.append(node.val)

"""
Non-Recursion
"""
class Solution_2:
    def postorderTraversal(self, root: TreeNode) -> List[int]:
        if not root:
            return []

        stack = [root]
        res = []
        while stack:
            root = stack.pop()
            res.append(root.val)
            if root.left:
                stack.append(root.left)
            if root.right:
                stack.append(root.right)

        return res[::-1]

File: leetcode/C++/0145-h-binary-tree-postorder-traversal.cpp
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> result;

        if (root == nullptr)
            return result;
            
        stack<const TreeNode *> s;
        const TreeNode *p = root;
        s.push(p);

        while (!s.empty()) {
            p = s.top();
            s.pop();
            result.push_back(p->val);
            if (p->left != nullptr)
                s.push(p->left);
            if (p->right != nullptr)
                s.push(p->right);
        }

        reverse(result.begin(), result.end());

        return result;
    }
};